### PR TITLE
Edge collapsing added and tested

### PR DESCRIPTION
**Pull request by Yixin-Hu**
1. add edge collapsing with the new timestamp
2. partially added consolidate_mesh() function

### CODE REVIEW

File: app/tetwild/TetWild.cpp [hunk 1]
```diff
@@ -150,8 +150,58 @@ bool tetwild::TetWild::smooth_after(const Tuple& t)
...
+//    for (size_t i = 0; i < m_tet_attribute.size(); i++) {
+//        if (m_tet_attribute[i].m_is_removed) continue;
+//        new_m_tet_attribute[t_cnt] = m_tet_attribute[i];
+//        t_cnt++;
+//    }
+//
+//    m_vertex_attribute = new_m_vertex_attribute;
+//    m_tet_attribute = new_m_tet_attribute;
+//
+//}
```
Comment (jiangzhongshi) [inline app/tetwild/TetWild.cpp:193]:
```suggestion
//void tetwild::TetWild::consolidate_mesh_attributes(){
//    int v_cnt = 0;
//    std::vector<size_t> map_v_ids(m_vertex_attribute.size(), -1);
//    for (size_t i = 0; i < m_vertex_attribute.size(); i++) {
//        if (m_vertex_attribute[i].m_is_removed) continue;
//        map_v_ids[i] = v_cnt;
//        v_cnt++;
//    }
//    int t_cnt = 0;
//    std::vector<size_t> map_t_ids(m_tet_attribute.size(), -1);
//    for (size_t i = 0; i < m_tet_attribute.size(); i++) {
//        if (m_tet_attribute[i].m_is_removed) continue;
//        map_t_ids[i] = t_cnt;
//        t_cnt++;
//    }
//
//#ifdef WILDMESHING_TOOLKIT_WITH_TBB
//    tbb::concurrent_vector<VertexAttributes> new_m_vertex_attribute(v_cnt);
//    tbb::concurrent_vector<TetAttributes> new_m_tet_attribute(t_cnt);
//#else
//    std::vector<VertexAttributes> new_m_vertex_attribute(v_cnt);
//    std::vector<TetAttributes> new_m_tet_attribute(t_cnt);
//#endif
//
//    v_cnt = 0;
//    for (size_t i = 0; i < m_vertex_attribute.size(); i++) {
//        if (m_vertex_attribute[i].m_is_removed) continue;
//        new_m_vertex_attribute[v_cnt] = m_vertex_attribute[i];
//        v_cnt++;
//    }
//    t_cnt = 0;
//    for (size_t i = 0; i < m_tet_attribute.size(); i++) {
//        if (m_tet_attribute[i].m_is_removed) continue;
//        new_m_tet_attribute[t_cnt] = m_tet_attribute[i];
//        t_cnt++;
//    }
//
//    m_vertex_attribute = new_m_vertex_attribute;
//    m_tet_attribute = new_m_tet_attribute;
//
//}
```

File: src/EdgeCollapsingConn.cpp [hunk 2]
```diff
@@ -5,9 +5,146 @@
...
+        int j = m_tet_connectivity[t_id].find(v1_id);
+        assert(j >= 0);
+        m_tet_connectivity[t_id][j] = v2_id;
+    }
+    vector_sort(m_vertex_connectivity[v2_id].m_conn_tets);
+    //
+    m_vertex_connectivity[v1_id].m_conn_tets.clear(); // release mem
+
+    Tuple new_loc = tuple_from_vertex(v2_id);
+    // todo: update timestamp of tets and tuple, change locs to loc
```
Comment (jiangzhongshi) [inline src/EdgeCollapsingConn.cpp:70]: This is stll todo?

File: src/EdgeCollapsingConn.cpp [hunk 3]
```diff
@@ -5,9 +5,146 @@
...
+        //
+        //        assert(old_m_vertex_connectivity == m_vertex_connectivity);
+        //        assert(old_m_tet_connectivity == m_tet_connectivity);
+        //        //fortest
+
+        return false;
+    }
+
+    /// call invariants on all entities
+    if (false) // if any invariant fails
```
Comment (jiangzhongshi) [inline src/EdgeCollapsingConn.cpp:115]: why `if(false)`

File: src/EdgeSplittingConn.cpp [hunk 4]
```diff
@@ -84,21 +84,23 @@ bool wmtk::TetMesh::split_edge(const Tuple& loc0, std::vector<Tuple>& new_edges)
         }
     }
 
-    // possibly call the resize_attributes
+    /// checks (possibly call the resize_attributes
```
Comment (jiangzhongshi) [inline src/EdgeSplittingConn.cpp:87]: Why triple-slash everywhere?

File: src/EdgeCollapsingConn.cpp (hunk 2 above)
Comment (Yixin-Hu) [reply src/EdgeCollapsingConn.cpp:70]: I'll change it

File: src/EdgeCollapsingConn.cpp (hunk 3 above)
Comment (Yixin-Hu) [reply src/EdgeCollapsingConn.cpp:115]: invariant checks which probably the same as the link condition.

File: src/TetMesh.h [hunk 5]
```diff
@@ -246,6 +262,21 @@ class TetMesh
...
+
+        friend bool operator==(const TetrahedronConnectivity& l, const TetrahedronConnectivity& r)
+        {
+            return std::tie(l.m_indices, l.m_is_removed, l.timestamp) ==
+                   std::tie(r.m_indices, r.m_is_removed, r.timestamp); // keep the same order
+        }
+
+        void print_info()
+        {
+            std::cout << "m_indices: ";
```
Comment (jiangzhongshi) [inline src/TetMesh.h:274]: Please replace with spdlog or not putting in the general header.

File: app/tetwild/TetWild.h [hunk 6]
```diff
@@ -99,23 +99,40 @@ class TetWild : public wmtk::TetMesh
...
     bool is_inverted(const Tuple& loc);
     double get_quality(const Tuple& loc);
 
     bool vertex_invariant(const Tuple& t) override;
     bool tetrahedron_invariant(const Tuple& t) override;
+
+    void consolidate_mesh();
+    //    void consolidate_mesh_attributes();
+
+    void pausee()
```
Comment (jiangzhongshi) [inline app/tetwild/TetWild.h:129]: Maybe put it into utils? not in tetwild class?

**Pull request merged**